### PR TITLE
`RichText` - A drop-in replacement for `TText` allowing you to render Markdown and LaTeX

### PR DESCRIPTION
**Pull request by holoisme**
Hello,

Up until now, our application textual content was rendered in the most literal way, where each character is mapped to a font glyph. This is amazingly fast and efficient of course, but quite limiting when we want to render more complex and unpredictable user-generated content, such as mathematical expressions, code or titles.

### Introducting `RichText`
This PR introduces a the new `RichText` component which can handle :
 - Inline mathematical expressions (`$a+b$`)
 - Centered mathematical expressions (`$$\sum x_i$$`)
 - Code
 - Multi-level titles (`# My title here`)

The supported programming languages are JavaScript, Typescript, C, Python, Java, Fox :fox_face:  and Rust :heart:, but it will be very easy to add more in the future as I designed the lexer to be very flexible and capable to handle special syntax.

Example usage:
```typescript
<RichText px={'sm'} py={'md'}>
  {`
  # Euler's identity
  Considered the most beautiful mathematical 
  it combines all fundamental mathematical constants
  of our universe.
  The most common form is $$e^{i\\pi} + 1 = 0$$
  #### Constants
  $e$ is Euler's number, the base of natural logarithms,
  $i$ is the imaginary unit
  $\\pi$ is the ratio of the circumference of a circle to its diameter.
  $1$ is the multiplicative unit
  $0$ is the additive unit
  `}
</RichText>
```

image

image

image

### Warning
Don't abuse it. `RichText` uses a web view to render custom `HTML` syntax so it might take 1 to 3 frames to render, which provoke a small initial flicker. Use it only with **user (student/professor) generated content** where it makes sense to use it :+1:.

### CODE REVIEW

File: edweiss-app/components/core/rich-text/RichText.tsx
Comment (FlorianDinant) [inline edweiss-app/components/core/rich-text/RichText.tsx:1]: This is a very good code ! Everything is well structured and well written. Congrats for that!!! I suggest you to add some documentation to make it more understandable but overall it is really great! 

Also, do you think it is ok to put everything on a same file? Don't you think we could divise it into multiple files in a `utils/richtext/` directory or anything similar ? There is a lot of constants and functions on a same file.

**Review by FlorianDinant — CHANGES_REQUESTED** [review]
Overall it is a very good PR! I don't have a lot to say on the code logic, we clearly see that you know what you are doing and which functions/hooks to use. However you could improve documentation and structure for readability and understanding in some parts.

**Comment by holoisme** [comment]
> Overall it is a very good PR! I don't have a lot to say on the code logic, we clearly see that you know what you are doing and which functions/hooks to use. However you could improve documentation and structure for readability and understanding in some parts.
> 
> Also, I see that there is many new issues on **`sonarcloud`**. Can you look at it and try to fix them?

Hello, thank you for your review.
Sonarcloud confirmed once again that it is terrible at identifying mistakes/bugs, for example take a look at the very first issue :

image

It says that I should use the method `.endsWith()` of `String`, because it sees as `$` at the end of the Regex expression and thinks I'm referring to an end of line, not seeing that I escaped it with a `\` and that I actually refer to the actual character `$`.

In general, do not use Sonarcloud's issues as a metric for code quality, especially for non-trivial code, don't take them seriously and if you see real issues let's discuss :)

File: edweiss-app/__test__/components/core/rich-text/__snapshots__/RichText.test.ts.snap [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`RichText should parse code and comments correctly 1`] = `
+"
+			<div style="
```
[the remark this answers is not in the document]
Comment (holoisme) [reply edweiss-app/__test__/components/core/rich-text/__snapshots__/RichText.test.ts.snap:5]: Unfortunately it's the nature of what the function is doing, it parses things and turns them into very compact HTML. (for webview parsing performance).

File: edweiss-app/components/core/rich-text/RichText.tsx [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+
+	return total;
+}
+
+function headerLevel(line: string, level: number) {
+	return `<div style="font-size: ${headerSizes[level - 1]}px; font-weight: bold">${line.substring(level + 1)}</div>`;
+}
+
+export function richTextToHTML(rawText: string, theme: Theme, options?: RichTextOptions): string {
+	// const text = sanitizeHtml(rawText, {
```
[the remark this answers is not in the document]
Comment (holoisme) [reply edweiss-app/components/core/rich-text/RichText.tsx:405]: You are right, this should be removed.

File: edweiss-app/components/core/rich-text/WebViewMathJaxWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+				cacheMode='LOAD_CACHE_ELSE_NETWORK'
+				{...props.webview}
+			/>
+		</View>
+	);
+};
+
+export default WebViewMathJaxWrapper;
+
+// const [mathJax, setMathJax] = useState<string | undefined>(cachedMathJax);
```
[the remark this answers is not in the document]
Comment (holoisme) [reply edweiss-app/components/core/rich-text/WebViewMathJaxWrapper.tsx:67]: They are a tentative to cache the `MathJax.js` that I was hoping to pursue in the future, but I guess it's not necessary so I should remove them.

File: edweiss-app/components/core/rich-text/RichText.tsx
Comment (holoisme) [reply edweiss-app/components/core/rich-text/RichText.tsx:1]: You're right, I created a new file `constants/RichTextSpecifications.ts` which hold all programming languages specifications.

**Review by FlorianDinant — APPROVED** [review]
It seems everything is good now. Good job! You have my green flag to merge it into main.

**Pull request merged**